From aa94239236e9bb27f6e844a0aa3941eac45a3fa1 Mon Sep 17 00:00:00 2001
From: Andi Kleen <andi@firstfloor.org>
Date: Sun, 2 May 2021 06:49:51 -0700
Subject: [PATCH] btf: Fix up btf_ids.h for gcc lto

gcc lto doesn't guarantee that top level asm() statements are in a
particular order, and also they might end up in different assembler
files.

This causes problems with the way the btf ids are generated, which
assumes a hole less placement of asm statements.

Make all the symbols global so that we avoid linker errors
if the asm statement is somewhere else.

For BTF list use a single asm() statement through the macros,
to keep them all together.

Probably it would be better to rewrite some of this in C, then
we could at least use __noreorder.

Cc: jolsa@kernel.org
Signed-off-by: Andi Kleen <andi@firstfloor.org>
---
 include/linux/btf_ids.h | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/include/linux/btf_ids.h b/include/linux/btf_ids.h
index 57890b357f85..69fabb8ceb03 100644
--- a/include/linux/btf_ids.h
+++ b/include/linux/btf_ids.h
@@ -27,7 +27,7 @@ struct btf_id_set {
 #define ____BTF_ID(symbol)				\
 asm(							\
 ".pushsection " BTF_IDS_SECTION ",\"a\";       \n"	\
-".local " #symbol " ;                          \n"	\
+".globl " #symbol " ;                          \n"	\
 ".type  " #symbol ", STT_OBJECT;               \n"	\
 ".size  " #symbol ", 4;                        \n"	\
 #symbol ":                                     \n"	\
@@ -70,7 +70,7 @@ asm(							\
 ".popsection;                                  \n");
 
 #define BTF_ID_LIST(name)				\
-__BTF_ID_LIST(name, local)				\
+__BTF_ID_LIST(name, globl)				\
 extern u32 name[];
 
 #define BTF_ID_LIST_GLOBAL(name)			\
@@ -124,10 +124,10 @@ asm(							\
 "." #scope " __BTF_ID__set__" #name ";         \n"	\
 "__BTF_ID__set__" #name ":;                    \n"	\
 ".zero 4                                       \n"	\
-".popsection;                                  \n");
+".popsection;                                  \n"
 
 #define BTF_SET_START(name)				\
-__BTF_ID_LIST(name, local)				\
+__BTF_ID_LIST(name, globl)				\
 __BTF_SET_START(name, local)
 
 #define BTF_SET_START_GLOBAL(name)			\
@@ -135,7 +135,6 @@ __BTF_ID_LIST(name, globl)				\
 __BTF_SET_START(name, globl)
 
 #define BTF_SET_END(name)				\
-asm(							\
 ".pushsection " BTF_IDS_SECTION ",\"a\";      \n"	\
 ".size __BTF_ID__set__" #name ", .-" #name "  \n"	\
 ".popsection;                                 \n");	\
